app/lib/actions/remote_execution/run_host_job.rb

```diff
@@ -29,15 +29,16 @@ def plan(job_invocation, host, template_invocation, proxy_selector = ::RemoteExe
 
         raise _('Could not use any template used in the job invocation') if template_invocation.blank?
 
+        provider = template_invocation.template.provider
+        proxy_selector = provider.required_proxy_selector_for(template_invocation.template) || proxy_selector
```

Comment:
I think it'd be better to make the default in the class nil if something else is going to override it later. Otherwise it's a bit of an odd surprise to the API consumer.

Or the `ProxySelector` should have a way to filter based on what the template provided.

Reply:
This action is only planned from one place and the selector is explicitly provided, so the default is in fact never used. This is done so all sub-tasks in a batch share the same selector which keeps some local state. However, in here we need to override the selector on a per-template (and therefore per-host) basis so it felt wrong to make the decision which selector to use and pass down at the parent task level.

That's the reason why I ended up with a somewhat surprising interface where a function may choose not to use some of its arguments based on another argument.

I considered having `ProxySelector` filter based on a template but that didn't feel all that great at the time. I'd say the selector's job is to select a proxy, it shouldn't need to know anything about templates.

I'll sleep on it and see if I can come up with something better